presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/CrossJoinWithArrayNotContainsToAntiJoin.java

```diff
@@ -0,0 +1,224 @@
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static com.facebook.presto.SystemSessionProperties.isRewriteCrossJoinArrayNotContainsToAntiJoinEnabled;
+import static com.facebook.presto.common.type.BigintType.BIGINT;
+import static com.facebook.presto.common.type.BooleanType.BOOLEAN;
+import static com.facebook.presto.common.type.DateType.DATE;
+import static com.facebook.presto.common.type.IntegerType.INTEGER;
+import static com.facebook.presto.common.type.VarcharType.VARCHAR;
+import static com.facebook.presto.expressions.LogicalRowExpressions.and;
+import static com.facebook.presto.expressions.LogicalRowExpressions.extractConjuncts;
+import static com.facebook.presto.matching.Capture.newCapture;
+import static com.facebook.presto.spi.relation.SpecialFormExpression.Form.IS_NULL;
+import static com.facebook.presto.sql.planner.PlannerUtils.addProjections;
+import static com.facebook.presto.sql.planner.PlannerUtils.isValidArrayContainsFilter;
+import static com.facebook.presto.sql.planner.PlannerUtils.restrictOutput;
+import static com.facebook.presto.sql.planner.plan.Patterns.filter;
+import static com.facebook.presto.sql.planner.plan.Patterns.join;
+import static com.facebook.presto.sql.planner.plan.Patterns.source;
+import static com.facebook.presto.sql.planner.plan.Patterns.sources;
+import static com.facebook.presto.sql.relational.Expressions.call;
+import static com.facebook.presto.sql.relational.Expressions.specialForm;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Inner join with contains function inside join clause will be run as cross join with filter.
+ * When the join condition has pattern of contains(array, element), we can rewrite it to a inner join. For example:
+ * <pre>
+ * - Filter not contains(r_array, l_key)
+ *      - Cross join
+ *          - scan l
+ *          - scan r
+ * </pre>
+ * into:
+ * <pre>
+ *     - Filter (l_array_elem IS NULL)
+ *       - LOJ (l_key = l_array_elem)
+ *          - scan l
+ *          - Unnest
+ *              l_array_elem <- unnest distinct_array
+ *              - project
+ *                  distinct_array := array_distinct(r_array)
+ *                  - scan r
+ * </pre>
+ */
+public class CrossJoinWithArrayNotContainsToAntiJoin
+        implements Rule<FilterNode>
+{
+    private static final List<Type> SUPPORTED_JOIN_KEY_TYPES = ImmutableList.of(BIGINT, INTEGER, VARCHAR, DATE);
+
+    private static final Capture<JoinNode> JOIN = newCapture();
+
+    private static final Capture<List<PlanNode>> JOIN_CHILDREN = Capture.newCapture();
+
+    private static final Pattern<FilterNode> PATTERN = filter()
+            .with(source().matching(join().matching(x -> x.getType().equals(JoinNode.Type.INNER) && x.getCriteria().isEmpty()).capturedAs(JOIN).with(sources().capturedAs(JOIN_CHILDREN))));
```

Comment:
use https://github.com/prestodb/presto/blob/833ecd0c62d3fdcd4be90620edd2d18dbda41eb9/presto-main/src/main/java/com/facebook/presto/sql/planner/plan/JoinNode.java#L374 to check?